File: java/com/joshmanisdabomb/lcc/entity/AtomicBombEntity.java
package com.joshmanisdabomb.lcc.entity;

import com.joshmanisdabomb.lcc.block.AtomicBombBlock;
import com.joshmanisdabomb.lcc.data.capability.NuclearCapability;
import com.joshmanisdabomb.lcc.registry.LCCBlocks;
import com.joshmanisdabomb.lcc.registry.LCCEntities;
import com.joshmanisdabomb.lcc.tileentity.AtomicBombTileEntity;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.MoverType;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.DirectionalPlaceContext;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ShearsItem;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.network.IPacket;
import net.minecraft.network.PacketBuffer;
import net.minecraft.network.datasync.DataParameter;
import net.minecraft.network.datasync.DataSerializers;
import net.minecraft.network.datasync.EntityDataManager;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.Direction;
import net.minecraft.util.Hand;
import net.minecraft.util.SoundEvents;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.GameRules;
import net.minecraft.world.World;
import net.minecraftforge.common.util.Constants;
import net.minecraftforge.fml.common.registry.IEntityAdditionalSpawnData;
import net.minecraftforge.items.ItemStackHandler;

import javax.annotation.Nullable;

public class AtomicBombEntity extends Entity implements LCCEntityHelper, IEntityAdditionalSpawnData {

    private static final DataParameter<Boolean> ACTIVE = EntityDataManager.createKey(AtomicBombEntity.class, DataSerializers.BOOLEAN);
    private static final DataParameter<Integer> FUSE = EntityDataManager.createKey(AtomicBombEntity.class, DataSerializers.VARINT);
    private static final DataParameter<Direction> FACING = EntityDataManager.createKey(AtomicBombEntity.class, DataSerializers.DIRECTION);
    public CompoundNBT tileEntityData;

    protected LivingEntity tntPlacedBy;

    private boolean active = false;
    private int fuse = 1200;
    public Direction facing = Direction.NORTH;

    private AxisAlignedBB bb = super.getBoundingBox();
    private int fallTime = 0;

    public AtomicBombEntity(EntityType<AtomicBombEntity> type, World world) {
        super(type, world);
    }

    public AtomicBombEntity(World world, double x, double y, double z, Direction facing, AtomicBombTileEntity te) {
        this(LCCEntities.atomic_bomb, world);
        this.setDirection(facing);
        this.bb = super.getBoundingBox().offset(x, y, z);
        this.setPosition(x, y, z);
        this.setMotion(Vec3d.ZERO);
        this.tileEntityData = te.write(new CompoundNBT());
        this.prevPosX = x;
        this.prevPosY = y;
        this.prevPosZ = z;
    }

    public AtomicBombEntity(World world, double x, double y, double z, Direction facing, AtomicBombTileEntity te, @Nullable LivingEntity entity) {
        this(world, x, y, z, facing, te);
        double lvt_9_1_ = world.rand.nextDouble() * 6.2831854820251465D;
        this.setMotion(-Math.sin(lvt_9_1_) * 0.02D, 0.20000000298023224D, -Math.cos(lvt_9_1_) * 0.02D);
        this.setActive(true);
        this.setFuse(NuclearCapability.getFuse(this.getUranium()));
        this.tntPlacedBy = entity;
    }

    @Override
    public IPacket<?> createSpawnPacket() {
        return this.traitCreateSpawnPacket();
    }

    protected void registerData() {
        this.dataManager.register(FUSE, 1200);
        this.dataManager.register(FACING, Direction.NORTH);
        this.dataManager.register(ACTIVE, false);
    }

    protected boolean canTriggerWalking() {
        return false;
    }

    public boolean canBeCollidedWith() {
        return this.isAlive();
    }

    public void tick() {
        this.prevPosX = this.getPosX();
        this.prevPosY = this.getPosY();
        this.prevPosZ = this.getPosZ();

        if (!this.hasNoGravity()) {
            this.setMotion(this.getMotion().add(0.0D, -0.08D, 0.0D));
        }

        this.move(MoverType.SELF, this.getMotion());
        this.setMotion(this.getMotion().scale(0.98D));
        if (this.onGround) {
            this.setMotion(this.getMotion().mul(0.12D, -0.5D, 0.12D));
        }

        if (this.active) {
            --this.fuse;
            if (this.fuse <= 0) {
                this.remove();
                if (!this.world.isRemote) {
                    this.explode();
                }
            } else {
                this.handleWaterMovement();
            }
        } else {
            if (!this.world.isRemote) {
                BlockPos bp = new BlockPos(this);
                if (this.onGround) {
                    boolean movingPiston = false;
                    boolean replaceable = true;
                    boolean setBlockState;
                    for (int i = -1; i <= 1; i++) {
                        BlockPos bp2 = bp.offset(this.facing, i);
                        BlockState state = this.world.getBlockState(bp);
                        movingPiston = movingPiston || state.getBlock() == Blocks.MOVING_PISTON;
                        replaceable = replaceable && state.isReplaceable(new DirectionalPlaceContext(this.world, bp, Direction.DOWN, ItemStack.EMPTY, Direction.UP));
                    }
                    if (!movingPiston) {
                        if (replaceable) {
                            BlockState newState = LCCBlocks.atomic_bomb.getDefaultState().with(AtomicBombBlock.FACING, this.facing).with(AtomicBombBlock.SEGMENT, AtomicBombBlock.Segment.MIDDLE);
                            setBlockState = this.world.setBlockState(bp, newState, 3);
                            setBlockState = setBlockState && this.world.setBlockState(bp.offset(this.facing), newState.with(AtomicBombBlock.SEGMENT, AtomicBombBlock.Segment.FRONT), 3);
                            setBlockState = setBlockState && this.world.setBlockState(bp.offset(this.facing.getOpposite()), newState.with(AtomicBombBlock.SEGMENT, AtomicBombBlock.Segment.BACK), 3);
                            if (setBlockState) {
                                if (this.tileEntityData != null) {
                                    TileEntity te = this.world.getTileEntity(bp);
                                    if (te != null) {
                                        CompoundNBT data = te.write(new CompoundNBT()).merge(this.tileEntityData);
                                        data.putInt("x", bp.getX());
                                        data.putInt("y", bp.getY());
                                        data.putInt("z", bp.getZ());
                                        te.read(data);
                                        te.markDirty();
                                    }
                                }
                                this.remove();
                            } else {
                                this.drops();
                                this.remove();
                            }
                        } else {
                            this.drops();
                            this.remove();
                        }
                    }
                } else {
                    if ((this.fallTime > 100 && (bp.getY() < 1 || bp.getY() > 256)) || this.fallTime > 600) {
                        this.drops();
                        this.remove();
                    }
                }
                this.fallTime++;
            }
        }
    }

    private void drops() {
        if (this.world.getGameRules().getBoolean(GameRules.DO_ENTITY_DROPS)) {
            this.entityDropItem(LCCBlocks.atomic_bomb);
            if (this.tileEntityData != null && this.tileEntityData.contains("inventory", Constants.NBT.TAG_COMPOUND)) {
                ItemStackHandler h = new ItemStackHandler();
                h.deserializeNBT(this.tileEntityData.getCompound("inventory"));
                for (int i = 0; i < h.getSlots(); i++) {
                    this.entityDropItem(h.extractItem(i, 64, false));
                }
            }
        }
    }

    protected void explode() {
        world.addEntity(new NuclearExplosionEntity(this.world, this.getPosX(), this.getPosY(), this.getPosZ(), (short)(NuclearCapability.getExplosionLifetime(this.getUranium(), false))));
    }

    private int getUranium() {
        int u = 0;
        if (this.tileEntityData != null && this.tileEntityData.contains("inventory", Constants.NBT.TAG_COMPOUND)) {
            ItemStackHandler h = new ItemStackHandler();
            h.deserializeNBT(this.tileEntityData.getCompound("inventory"));
            for (int i = 0; i < h.getSlots(); i++) {
                ItemStack stack = h.getStackInSlot(i);
                if (stack.getItem() == LCCBlocks.enriched_uranium_storage.asItem()) u += (stack.getCount() * 9);
            }
        }
        return u;
    }

    @Override
    public boolean processInitialInteract(PlayerEntity player, Hand hand) {
        if (this.active) {
            ItemStack itemstack = player.getHeldItem(hand);
            if (itemstack.getItem() instanceof ShearsItem) {
                this.world.playSound(player, this.getPosX(), this.getPosY(), this.getPosZ(), SoundEvents.ENTITY_SHEEP_SHEAR, this.getSoundCategory(), 1.0F, 1.0F);
                player.swingArm(hand);
                if (!this.world.isRemote) {
                    this.drops();
                    this.remove();
                    itemstack.damageItem(1, player, (p_213625_1_) -> {
                        p_213625_1_.sendBreakAnimation(hand);
                    });
                    return true;
                }
            }
        }
        return super.processInitialInteract(player, hand);
    }

    @Override
    protected void readAdditional(CompoundNBT compound) {
        this.setFuse(compound.getShort("Fuse"));
        this.setActive(compound.getBoolean("Active"));
        this.setDirection(Direction.byHorizontalIndex(compound.getByte("Facing")));
        this.fallTime = compound.getInt("Fall");
        if (compound.contains("TileEntityData", 10)) {
            this.tileEntityData = compound.getCompound("TileEntityData");
        }
    }

    @Override
    protected void writeAdditional(CompoundNBT compound) {
        compound.putShort("Fuse", (short)this.getFuse());
        compound.putBoolean("Active", this.isActive());
        compound.putByte("Facing", (byte)this.facing.getHorizontalIndex());
        compound.putInt("Fall", this.fallTime);
        if (this.tileEntityData != null) {
            compound.put("TileEntityData", this.tileEntityData);
        }
    }

    @Override
    public void notifyDataManagerChange(DataParameter<?> key) {
        if (FUSE.equals(key)) this.fuse = this.dataManager.get(FUSE);
        if (ACTIVE.equals(key)) this.active = this.dataManager.get(ACTIVE);
        if (FACING.equals(key)) {
            this.facing = this.dataManager.get(FACING);
            this.recalculateSize();
        }
    }

    public int getFuse() {
        return this.fuse;
    }

    public boolean isActive() {
        return this.active;
    }

    public void setFuse(int fuseIn) {
        this.dataManager.set(FUSE, this.fuse = fuseIn);
    }

    public void setActive(boolean active) {
        this.dataManager.set(ACTIVE, this.active = active);
    }

    public void setDirection(Direction facing) {
        this.dataManager.set(FACING, this.facing = facing);
    }

    @Override
    public AxisAlignedBB getBoundingBox() {
        return this.bb;
    }

    @Override
    public void setBoundingBox(AxisAlignedBB b) {
        this.bb = b;
        super.setBoundingBox(b);
    }

    @Override
    public void recalculateSize() {
        super.recalculateSize();
        this.setBoundingBox(this.expandBoundingBox(this.getBoundingBox()));
    }

    @Override
    public void setPosition(double x, double y, double z) {
        super.setPosition(x, y, z);
        this.setBoundingBox(this.expandBoundingBox(this.getBoundingBox()));
    }

    private AxisAlignedBB expandBoundingBox(AxisAlignedBB b) {
        if (facing != null) return b.grow(Math.abs(facing.getXOffset()), 0, Math.abs(facing.getZOffset()));
        return b;
    }

    @Override
    public void writeSpawnData(PacketBuffer buffer) {
        buffer.writeEnumValue(this.facing);
    }

    @Override
    public void readSpawnData(PacketBuffer buffer) {
        this.bb = super.getBoundingBox().offset(this.getPosX(), this.getPosY(), this.getPosZ());
        this.setDirection(buffer.readEnumValue(Direction.class));
        this.recalculateSize();
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/entity/AtomicBombEntity.kt
package com.joshmanisdabomb.lcc.entity

import com.joshmanisdabomb.lcc.abstracts.nuclear.NuclearUtil
import com.joshmanisdabomb.lcc.block.AtomicBombBlock
import com.joshmanisdabomb.lcc.block.entity.AtomicBombBlockEntity
import com.joshmanisdabomb.lcc.directory.LCCBlocks
import com.joshmanisdabomb.lcc.directory.LCCChunkTickets
import com.joshmanisdabomb.lcc.directory.LCCEntities
import com.joshmanisdabomb.lcc.directory.LCCSounds
import com.joshmanisdabomb.lcc.extensions.*
import com.joshmanisdabomb.lcc.inventory.AtomicBombInventory
import com.joshmanisdabomb.lcc.trait.LCCEntityTrait
import net.fabricmc.fabric.api.tool.attribute.v1.FabricToolTags
import net.minecraft.block.Blocks
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityType
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.MovementType
import net.minecraft.entity.data.DataTracker
import net.minecraft.entity.data.TrackedData
import net.minecraft.entity.data.TrackedDataHandlerRegistry
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.Inventories
import net.minecraft.item.AutomaticItemPlacementContext
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.nbt.NbtCompound
import net.minecraft.server.world.ServerWorld
import net.minecraft.sound.SoundCategory
import net.minecraft.state.property.Properties
import net.minecraft.text.Text
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.Unit
import net.minecraft.util.math.Box
import net.minecraft.util.math.ChunkPos
import net.minecraft.util.math.Direction
import net.minecraft.util.math.Vec3d
import net.minecraft.world.GameRules
import net.minecraft.world.World
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin

class AtomicBombEntity(type: EntityType<*>, world: World) : Entity(type, world), LCCEntityTrait {

    constructor(world: World, x: Double, y: Double, z: Double, facing: Direction, be: AtomicBombBlockEntity?) : this(LCCEntities.atomic_bomb, world) {
        facing.also { _facing = it; this.facing = it }
        updatePosition(x, y, z)
        velocity = Vec3d.ZERO
        this.be = be?.let {
            val nbt = NbtCompound()
            it.writeNbt(nbt)
            nbt
        }
        prevX = x
        prevY = y
        prevZ = z
    }

    constructor(world: World, x: Double, y: Double, z: Double, facing: Direction, be: AtomicBombBlockEntity?, igniter: LivingEntity?) : this(world, x, y, z, facing, be) {
        updatePosition(x, y, z)
        val d = world.random.nextDouble() * 6.2831854820251465
        this.setVelocity(-sin(d) * 0.02, 0.20000000298023224, -cos(d) * 0.02)
        true.also { _active = it; this.active = it }
        placedBy = igniter
    }

    var be: NbtCompound? = null

    var placedBy: LivingEntity? = null

    private var _active = false
    private var _fuse = 1200
    private var _facing = Direction.NORTH

    var active by active_data
    var fuse by fuse_data
    var facing by facing_data

    private var fallTime = 0

    val inventory get() = be?.let { AtomicBombInventory().apply { Inventories.readNbt(it, list) } }

    override fun initDataTracker() {
        dataTracker.startTracking(active_data, false)
        dataTracker.startTracking(fuse_data, 1200)
        dataTracker.startTracking(facing_data, Direction.NORTH)
    }

    override fun onTrackedDataSet(data: TrackedData<*>) {
        if (data == active_data) _active = dataTracker.get(active_data)
        if (data == fuse_data) _fuse = dataTracker.get(fuse_data)
        if (data == facing_data) {
            _facing = dataTracker.get(facing_data)
            calculateDimensions()
        }
    }

    override fun readCustomDataFromNbt(tag: NbtCompound) {
        tag.getBoolean("Active").apply { _active = this; active = this }
        tag.getShort("Fuse").run { if (this > 0) this else 1200 }.toInt().apply { _fuse = this; fuse = this }
        tag.getByte("Facing").run { Direction.fromHorizontal(this.toInt()) }.apply { _facing = this; facing = this }
        tag.getShort("Fall").toInt().apply { fallTime = this }
        if (tag.contains("TileEntityData", NBT_COMPOUND)) tag.getCompound("TileEntityData").apply { be = this }
    }

    override fun writeCustomDataToNbt(tag: NbtCompound) {
        tag.putBoolean("Active", _active)
        tag.putShort("Fuse", _fuse.toShort())
        tag.putByte("Facing", _facing.horizontal.toByte())
        tag.putShort("Fall", fallTime.toShort())
        be?.also { tag.put("TileEntityData", it) }
    }

    override fun createSpawnPacket() = lcc_createSpawnPacket()// { it.writeByte(_facing.horizontal) }

    override fun tick() {
        prevX = x
        prevY = y
        prevZ = z
        if (!hasNoGravity()) velocity = velocity.add(0.0, -0.08, 0.0)
        move(MovementType.SELF, velocity)
        velocity = velocity.multiply(0.98)
        if (onGround) velocity = velocity.multiply(0.12, -0.5, 0.12)
        if (_active) {
            if (!world.isClient) {
                for (i in -1..1) {
                    for (k in -1..1) {
                        if (i != 0 && k != 0) {
                            (world as ServerWorld).chunkManager.removeTicket(LCCChunkTickets.nuclear, ChunkPos(chunkPos.x + i, chunkPos.z + k), 17, Unit.INSTANCE)
                        }
                    }
                }
                (world as ServerWorld).chunkManager.addTicket(LCCChunkTickets.nuclear, chunkPos, 17, Unit.INSTANCE)
            }
            --fuse
            if (fuse <= 0) {
                discard()
                if (!world.isClient) inventory?.also { explode(it.uraniumCount) }
            } else {
                this.updateWaterState()
                if (fuse % 20 == 19) {
                    if (!world.isClient) world.playSound(null, pos.x + 0.5, pos.y + 0.5, pos.z + 0.5, LCCSounds.atomic_bomb_fuse, SoundCategory.BLOCKS, 5.0F, 1.2F - fuse.div(1200f).times(0.2f))
                }
            }
        } else {
            if (!world.isClient) {
                val bp = blockPos
                if (onGround) {
                    var movingPiston = false
                    var replaceable = true
                    var setBlockState: Boolean
                    for (i in -1..1) {
                        val bp2 = bp.offset(_facing, i)
                        val state = world.getBlockState(bp)
                        movingPiston = movingPiston || state.block === Blocks.MOVING_PISTON
                        replaceable = replaceable && state.canReplace(AutomaticItemPlacementContext(world, bp2, Direction.DOWN, ItemStack.EMPTY, Direction.UP))
                    }
                    if (!movingPiston) {
                        if (replaceable) {
                            val newState = LCCBlocks.atomic_bomb.defaultState.with(Properties.HORIZONTAL_FACING, _facing)
                            setBlockState = world.setBlockState(bp, newState, 3)
                            setBlockState = setBlockState && world.setBlockState(bp.offset(_facing), newState.with(AtomicBombBlock.segment, AtomicBombBlock.AtomicBombSegment.HEAD), 3)
                            setBlockState = setBlockState && world.setBlockState(bp.offset(_facing.opposite), newState.with(AtomicBombBlock.segment, AtomicBombBlock.AtomicBombSegment.TAIL), 3)
                            if (setBlockState) {
                                be?.also {
                                    (world.getBlockEntity(bp) as? AtomicBombBlockEntity)?.apply {
                                        val nbt = NbtCompound()
                                        this.writeNbt(nbt)
                                        val data = nbt.copyFrom(it)
                                        data.putInt("x", bp.x)
                                        data.putInt("y", bp.y)
                                        data.putInt("z", bp.z)
                                        this.readNbt(data)
                                        this.markDirty()
                                    }
                                }
                                this.discard()
                            } else {
                                this.spawnDrops()
                                this.discard()
                            }
                        } else {
                            this.spawnDrops()
                            this.discard()
                        }
                    }
                } else {
                    if (fallTime > 100 && (bp.y < 1 || bp.y > 256) || fallTime > 600) {
                        this.spawnDrops()
                        this.discard()
                    }
                }
                fallTime++
            }
        }
    }

    override fun interact(player: PlayerEntity, hand: Hand): ActionResult {
        if (this.active) {
            val stack = player.getStackInHand(hand)
            if (!player.isSurvival && stack.isOf(Items.FLINT_AND_STEEL)) {
                if (!world.isClient) explode(inventory?.uraniumCount ?: return ActionResult.PASS)
                return ActionResult.SUCCESS
            } else if (stack.isIn(FabricToolTags.SHEARS)) {
                world.playSound(null, pos.x + 0.5, pos.y + 0.5, pos.z + 0.5, LCCSounds.atomic_bomb_defuse, SoundCategory.BLOCKS, 1.0F, 1.0F)
                player.swingHand(hand)
                if (!world.isClient) {
                    spawnDrops()
                    discard()
                    stack.damage(1, player) { it.sendToolBreakStatus(hand) }
                    return ActionResult.SUCCESS
                }
            }
        }
        return ActionResult.PASS
    }

    private fun spawnDrops() {
        if (!world.gameRules.getBoolean(GameRules.DO_ENTITY_DROPS)) return

        dropItem(LCCBlocks.atomic_bomb)
        inventory?.forEach(this@AtomicBombEntity::dropStack)
    }

    private fun explode(uranium: Int) {
        NuclearExplosionEntity(world, x, y.plus(0.5), z, placedBy).also {
            it.radius = NuclearUtil.getExplosionRadiusFromUranium(uranium)
            it.lifetime = NuclearUtil.getExplosionLifetimeFromUranium(uranium)
            world.spawnEntity(it)
        }
        discard()
        (world as ServerWorld).chunkManager.addTicket(LCCChunkTickets.nuclear, chunkPos, 17, Unit.INSTANCE)
    }

    override fun calculateBoundingBox() = expandBoundingBox(super.calculateBoundingBox())

    private fun expandBoundingBox(default: Box): Box? {
        dataTracker.get(facing_data).apply { return default.expand(abs(offsetX.toDouble()), 0.0, abs(offsetZ.toDouble())) }
    }

    override fun isCollidable() = isAlive

    override fun collides() = isAlive

    override fun isPushable() = false

    override fun getName(): Text {
        customName?.apply { return this }
        be?.also {
            if (it.contains("CustomName", NBT_STRING)) return Text.Serializer.fromJson(it.getString("CustomName")) as Text
        }
        return super.getName()
    }

    override fun remove(reason: RemovalReason) {
        if (!world.isClient) (world as? ServerWorld)?.chunkManager?.removeTicket(LCCChunkTickets.nuclear, chunkPos, 17, Unit.INSTANCE)
        super.remove(reason)
    }

    companion object {
        val active_data = DataTracker.registerData(AtomicBombEntity::class.java, TrackedDataHandlerRegistry.BOOLEAN)
        val fuse_data = DataTracker.registerData(AtomicBombEntity::class.java, TrackedDataHandlerRegistry.INTEGER)
        val facing_data = DataTracker.registerData(AtomicBombEntity::class.java, TrackedDataHandlerRegistry.FACING)
    }

}